logproof/src/bfv_statement.rs

```diff
@@ -0,0 +1,861 @@
+    private_e: usize,
+    _phantom: PhantomData<(C, P)>,
+}
+
+impl<C, P> IdxOffsets<C, P> {
+    fn new(statements: &[BfvProofStatement<C, P>]) -> Self {
+        // Counts
+        let num_messages = Self::num_messages(statements);
+        let num_public = Self::num_public(statements);
+        let num_private = Self::num_private(statements);
+        let num_statements = statements.len();
+
+        // Offsets
+        let remainder = num_messages;
+        let public_key = remainder + num_statements;
+        let public_e_0 = public_key + num_public;
+        let public_e_1 = public_e_0 + num_public;
+        let private_a = public_e_1 + num_public;
+        let private_e = private_a + num_private;
+
+        Self {
+            remainder,
+            public_key,
+            public_e_0,
+            public_e_1,
+            private_a,
+            private_e,
+            _phantom: PhantomData,
+        }
+    }
+
+    /// Return the (row, col) shape of A.
+    fn a_shape(&self) -> (usize, usize) {
+        let num_private = self.private_e - self.private_a;
+        let num_public = self.public_e_0 - self.public_key;
+        (num_public * 2 + num_private, self.private_e + num_private)
+    }
+
+    /// Record that a private statement or witness has been inserted into `A` or `S`, respectively
+    /// bumping the indices.
+    fn inc_private(&mut self) {
+        self.remainder += 1;
+        self.private_a += 1;
+        self.private_e += 1;
+    }
+
+    /// Record that a public statement or witness has been inserted into `A` or `S`, respectively
+    /// bumping the indices.
+    fn inc_public(&mut self) {
+        self.remainder += 1;
+        self.public_key += 1;
+        self.public_e_0 += 1;
+        self.public_e_1 += 1;
+    }
+
+    fn num_messages(statements: &[BfvProofStatement<C, P>]) -> usize {
+        statements
+            .iter()
+            .fold(0usize, |max, s| usize::max(max, s.message_id()))
+            + 1
+    }
+
+    fn num_private(statements: &[BfvProofStatement<C, P>]) -> usize {
+        statements.iter().filter(|s| s.is_private()).count()
+    }
+
+    fn num_public(statements: &[BfvProofStatement<C, P>]) -> usize {
+        statements.len() - Self::num_private(statements)
+    }
+}
+
+trait AsPolynomial<R: Ring> {
+    fn as_poly(&self) -> Polynomial<R>;
+}
+
+trait AsPolynomials<R: Ring> {
+    fn as_poly_vec(&self) -> Vec<Polynomial<R>>;
+}
+
+impl<const N: usize, B: BarrettConfig<N>> AsPolynomial<Z<N, B>> for Plaintext {
+    fn as_poly(&self) -> Polynomial<Z<N, B>> {
+        Polynomial {
+            coeffs: strip_trailing_value(
+                (0..self.len())
+                    .map(|i| Zq::from(self.get_coefficient(i)))
+                    .collect::<Vec<_>>(),
+                Zq::zero(),
+            ),
+        }
+    }
+}
+
+impl<const N: usize, B: BarrettConfig<N>> AsPolynomial<Z<N, B>> for SecretKey {
+    fn as_poly(&self) -> Polynomial<Z<N, B>> {
+        todo!()
```

Comment:
Delete this impl or impl it.

Reply:
I'd prefer to leave it for now.. it's the only piece that needs to happen after we expose symmetric encryption from SEAL. If I remove the impl, I either need to comment out all of the private key encryption functionality in this module or do something nonsensical, like use a dummy polynomial, and I'd rather leave an explicit panic than something blatantly wrong. Would you like `unimplemented!()` better?